File: backend/src/models/Order.js
import { randomBytes } from 'crypto';
import mongoose from 'mongoose';

const orderItemSchema = new mongoose.Schema(
  {
    productId: { type: mongoose.Schema.Types.ObjectId, ref: 'Product' },
    name: String,
    price: Number,
    quantity: { type: Number, min: 1 },
    weightSnapshot: String,
    image: String,
  },
  { _id: false }
);

const orderSchema = new mongoose.Schema(
  {
    orderNumber: { type: String, unique: true },
    customerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      default: null,
    },
    guestName: { type: String, default: '' },
    guestPhone: { type: String, default: '' },
    items: [orderItemSchema],
    status: {
      type: String,
      enum: ['pending', 'confirmed', 'preparing', 'on_the_way', 'delivered', 'cancelled'],
      default: 'pending',
    },
    deliveryAddress: {
      label: String,
      line1: String,
      line2: String,
      city: String,
      region: String,
      postalCode: String,
      phone: String,
    },
    assignedDelivery: { type: String, default: '' },
    promoCode: { type: String, default: '' },
    discountAmount: { type: Number, default: 0 },
    subtotal: { type: Number, required: true },
    deliveryFee: { type: Number, default: 0 },
    total: { type: Number, required: true },
    notes: { type: String, default: '' },
    cancellationReason: { type: String, default: '' },
  },
  { timestamps: true }
);

orderSchema.index({ customerId: 1, createdAt: -1 });
orderSchema.index({ status: 1, createdAt: -1 });

orderSchema.pre('save', function (next) {
  if (!this.orderNumber) {
    const suffix = randomBytes(3).toString('hex').toUpperCase();
    this.orderNumber = `DTD-${Date.now()}-${suffix}`;
  }
  next();
});

export const Order = mongoose.model('Order', orderSchema);

File: mobile/lib/data/models/order.dart
class OrderItem {
  OrderItem({
    required this.productId,
    required this.name,
    required this.price,
    required this.quantity,
    this.weightSnapshot,
    this.image,
  });

  final String? productId;
  final String name;
  final double price;
  final int quantity;
  final String? weightSnapshot;
  final String? image;

  factory OrderItem.fromJson(Map<String, dynamic> j) {
    final pid = j['productId'];
    String? productId;
    if (pid is String) {
      productId = pid;
    } else if (pid is Map && pid['_id'] != null) {
      productId = pid['_id'].toString();
    }
    return OrderItem(
      productId: productId,
      name: j['name'] as String? ?? '',
      price: (j['price'] as num?)?.toDouble() ?? 0,
      quantity: (j['quantity'] as num?)?.toInt() ?? 1,
      weightSnapshot: j['weightSnapshot'] as String?,
      image: j['image'] as String?,
    );
  }
}

class Order {
  Order({
    required this.id,
    required this.orderNumber,
    required this.status,
    required this.items,
    required this.subtotal,
    required this.deliveryFee,
    required this.discountAmount,
    required this.total,
    required this.createdAt,
    this.assignedDelivery,
    this.deliveryAddress,
  });

  final String id;
  final String orderNumber;
  final String status;
  final List<OrderItem> items;
  final double subtotal;
  final double deliveryFee;
  final double discountAmount;
  final double total;
  final DateTime createdAt;
  final String? assignedDelivery;
  final Map<String, dynamic>? deliveryAddress;

  factory Order.fromJson(Map<String, dynamic> j) {
    final itemsRaw = j['items'] as List<dynamic>? ?? [];
    return Order(
      id: j['_id'] as String,
      orderNumber: j['orderNumber'] as String? ?? '',
      status: j['status'] as String? ?? 'pending',
      items: itemsRaw
          .map((e) => OrderItem.fromJson(e as Map<String, dynamic>))
          .toList(),
      subtotal: (j['subtotal'] as num?)?.toDouble() ?? 0,
      deliveryFee: (j['deliveryFee'] as num?)?.toDouble() ?? 0,
      discountAmount: (j['discountAmount'] as num?)?.toDouble() ?? 0,
      total: (j['total'] as num?)?.toDouble() ?? 0,
      createdAt: DateTime.parse(j['createdAt'] as String),
      assignedDelivery: j['assignedDelivery'] as String?,
      deliveryAddress: j['deliveryAddress'] as Map<String, dynamic>?,
    );
  }

  static String statusLabel(String s) {
    switch (s) {
      case 'confirmed':
        return 'Confirmed';
      case 'preparing':
        return 'Preparing';
      case 'on_the_way':
        return 'On the way';
      case 'delivered':
        return 'Delivered';
      case 'cancelled':
        return 'Cancelled';
      default:
        return 'Pending';
    }
  }
}
